src/backend/distributed/executor/executor.c

```diff
@@ -994,6 +1005,149 @@ FindOrCreateWorkerSession(WorkerPool *workerPool, MultiConnection *connection)
+ * a distributed deadlock when the upserts touch the same rows.
+ */
+static bool
+ShouldRunTasksSequentially(DistributedExecution *execution)
+{
+	DistributedPlan *plan = execution->plan;
+	Job *workerJob = plan->workerJob;
+	List *taskList = workerJob->taskList;
+	Task *initialTask = NULL;
+
+	if (list_length(taskList) < 2)
+	{
+		/* single task plans are already qualified as sequential by definition */
+		return false;
+	}
+
+	/* all the tasks are the same, so we only look one */
+	initialTask = (Task *) linitial(taskList);
+	if (initialTask->rowValuesLists != NIL)
+	{
+		/* found a multi-row INSERT */
+		return true;
+	}
+
+	return false;
+}
+
+
+/*
+ * SequentialRunDistributedExecution gets a distributed execution and
+ * executes each individual task in the exection sequentially, one
+ * task at a time. See related function ShouldRunTasksSequentially()
+ * for more detail on the definition of SequentialRun.
+ */
+static void
+SequentialRunDistributedExecution(DistributedExecution *execution)
+{
+	DistributedPlan *distributedPlan = execution->plan;
+	Job *job = distributedPlan->workerJob;
+	List *taskList = job->taskList;
+
+	int unfinishedTaskCount = execution->unfinishedTaskCount;
+	int taskIndex = 0;
+
+	Assert(list_length(taskList) == unfinishedTaskCount);
+
+	for (taskIndex = 0; taskIndex < unfinishedTaskCount; ++taskIndex)
+	{
+		DistributedExecution *singleTaskExecution = palloc0(sizeof(DistributedExecution));
+
+		Task *taskToExecute = (Task *) list_nth(taskList, taskIndex);
+		DistributedPlan *singleTaskDistributedPlan =
+			CreateSingleTaskDistributedPlan(taskToExecute, distributedPlan);
+
+		/*
+		 * Use some fields as the same which are safe to and required to
+		 * share across single task executions.
+		 */
+		singleTaskExecution->raiseInterrupts = execution->raiseInterrupts;
+		singleTaskExecution->isTransaction = execution->isTransaction;
+		singleTaskExecution->executionStats = execution->executionStats;
+		singleTaskExecution->scanState = execution->scanState;
+
+		/* we only expect to use a single connection for a single task */
+		singleTaskExecution->targetPoolSize = 1;
+		singleTaskExecution->unfinishedTaskCount = 1;
+		singleTaskExecution->totalTaskCount = 1;
+
+		/* RunDistributedExecution() will set the sessions, be explict here */
+		singleTaskExecution->sessionList = NIL;
+		singleTaskExecution->waitEventSet = NULL;
+
+		/* we're going to execute a single task only */
+		singleTaskExecution->plan = singleTaskDistributedPlan;
+
+		/*
+		 * StartDistributedExecution is common for all the executions (including
+		 * the parent execution) and so we don't need to re-call it for each
+		 * individual singleTaskExecution execution here.
+		 */
+
+		/* simply call the regular execution function */
+		RunDistributedExecution(singleTaskExecution);
+
+		/* remember the total number of rows processed */
+		execution->rowsProcessed += singleTaskExecution->rowsProcessed;
+
+		FinishDistributedExecution(singleTaskExecution);
```

Comment:
I'm still thinking whether this line is necessary/correct. Happy to get feedback on this.

Reply:
Some parts of `FinishDistributedExecution()` is necessary (e.g. most of the `foreach()` loop in there), and other parts seem harmless, so my take is that this is correct. I need to think more though.

Reply:
This does seem prone to bugs. If I'm making changes to these functions I don't expect create-start-create-run-finish-create-run-finish as a possible execution order.